Pegasus/include/Geometry.hpp

```diff
@@ -171,6 +171,397 @@ class Box : public SimpleShape
+* @brief Stores simplex size and current direction of search
+*/
+struct NearestSimplexData
+{
+    uint8_t simplexSize;
+    glm::dvec3 direction;
+};
+
+/**
+ * @brief Calculates whether a simplex contains origin and returns true if it does
+ *
+ * Can only be called when a simplex size is in the range of [2; 4], otherwise result is undefined
+ * @param[in] simplex simplex data
+ * @return @c true if a simplex contains origin @c false otherwise
+ */
+bool SimplexContainsOrigin(Simplex const& simplex);
+
+/**
+* @brief Calculates nearest simplex to the origin
+*
+* Presumes that simplex vertices are stored such that the latest added vertex has index @p simplexSize - 1
+* @param[in, out] simplex an array of vertices of a simplex
+* @param[in] simplexSize size of a simplex
+* @return new size of a simplex and a next directory of the search
+*/
+NearestSimplexData NearestSimplex(std::array<glm::dvec3, 4>& simplex, uint8_t simplexSize);
+
+/**
+ * @brief Checks if simplex contains origin and returs true if it does, 
+ * otherwise finds closest sub simplex to the origin
+ * @param[in, out] simplex current simplex
+ * @param[in, out] direction current search direction
+ * @return @c true if simplex contains origin, @c false otherwise
+ */
+bool DoSimplex(gjk::Simplex& simplex, glm::dvec3& direction);
+
+/**
+* @brief Attempts to calculate a tetrahedron from the CSO vertices such that it contains the origin
+* @tparam ShapeA any shape type for which gjk::Support is overloaded
+* @tparam ShapeB any shape type for which gjk::Support is overloaded
+* @param[in, out] simplex initial simplex
+* @param[in] aShape reference to the shape object
+* @param[in] bShape reference to the shape object
+* @param[in] direction initial search direction vector of unit length
+* @return @c true if there is intersection, @c false otherwise
+*/
+template <typename ShapeA, typename ShapeB>
+bool CalculateSimplex(Simplex& simplex, ShapeA const& aShape, ShapeB const& bShape, glm::dvec3 direction)
+{
+    do
+    {
+        //Add new vertex to the simplex
+        simplex.vertices[simplex.size++] = cso::Support(aShape, bShape, direction);
+
+        //Calculate if the new vertex is pass the origin
+        double const newPointDirectionProj = glm::dot(simplex.vertices[simplex.size - 1], direction);
+        if (math::fp::IsLess(newPointDirectionProj, 0.0))
+        {
+            return false;
+        }
+    } while (!DoSimplex(simplex, direction));
+
+    return true;
+}
+
+/**
+* @brief Calculates whether two shapes are intersecting using GJK algorithm
+* @tparam ShapeA any shape type for which gjk::Support is overloaded
+* @tparam ShapeB any shape type for which gjk::Support is overloaded
+* @param[in] aShape reference to the shape object
+* @param[in] bShape reference to the shape object
+* @return @c true if there is intersection, @c false otherwise
+*/
+template <typename ShapeA, typename ShapeB>
+bool CalculateIntersection(ShapeA const& aShape, ShapeB const& bShape)
```

Comment:
If you'd only need to use GJK without getting the resulting simplex you would use this one.